rethinkdb/rethinkdb.d.ts

```diff
@@ -8,229 +8,239 @@
-      not():Expression<boolean>;
-
-      gt(value:T):Expression<boolean>;
-      ge(value:T):Expression<boolean>;
-      lt(value:T):Expression<boolean>;
-      le(value:T):Expression<boolean>;
-
-      add(n:number):Expression<number>;
-      sub(n:number):Expression<number>;
-      mul(n:number):Expression<number>;
-      div(n:number):Expression<number>;
-      mod(n:number):Expression<number>;
-
-      hasFields(...fields:string[]):Expression<boolean>;
-
-      default(value:T):Expression<T>;
-  }
-
-  interface Operation<T> {
-   run(conn:Connection, cb?:(err:Error, result:T)=>void):Promise<T>;
-  }
-
-  interface Aggregator {}
-  interface Sort {}
-
-  interface Time {}
-
-
-  // http://www.rethinkdb.com/api/#js
-  // TODO control structures
+    export function connect(host: ConnectionOptions, cb?: (err: Error, conn: Connection) => void): Promise<Connection>;
+
+    export function dbCreate(name: string): Operation<CreateResult>;
+    export function dbDrop(name: string): Operation<DropResult>;
+    export function dbList(): Operation<string[]>;
+
+    export function db(name: string): Db;
+    export function table(name: string, options?: { useOutdated: boolean }): Table;
+
+    export function asc(property: string): Sort;
+    export function desc(property: string): Sort;
+
+    export var count: Aggregator;
+    export function sum(prop: string): Aggregator;
+    export function avg(prop: string): Aggregator;
+
+    export function row(name: string): Expression<any>;
+    export function expr(stuff: any): Expression<any>;
+
+    export function now(): Time;
+
+    // Control Structures
+    export function branch(test: Expression<boolean>, trueBranch: Expression<any>, falseBranch: Expression<any>): Expression<any>;
+
+
+    export class Cursor {
+        hasNext(): boolean;
+        each(cb: (err: Error, row: any) => void, done?: () => void): void;
+        each<T>(cb: (err: Error, row: T) => void, done?: () => void): void;
+        each(cb: (err: Error, row: any) => boolean, done?: () => void): void; // returning false stops iteration
+        each<T>(cb: (err: Error, row: T) => boolean, done?: () => void): void; // returning false stops iteration
+        next(cb: (err: Error, row: any) => void): void;
+        next<T>(cb: (err: Error, row: T) => void): void;
+        toArray(cb: (err: Error, rows: any[]) => void): void;
+        toArray<T>(cb: (err: Error, rows: T[]) => void): void;
+        toArray(): Promise<any[]>;
+        toArray<T>(): Promise<T[]>;
+        close(cb: (err: Error) => void): void;
+        close(): Promise<void>;
+    }
+
+    interface ConnectionOptions {
+        host: string;
+        port: number;
+        db?: string;
+        authKey?: string;
+    }
+
+    interface Connection {
+        close(cb: (err: Error) => void): void;
+        close(opts: { noreplyWait: boolean }, cb: (err: Error) => void): void;
+        close(): Promise<void>;
+        close(opts: { noreplyWait: boolean }): Promise<void>;
```

Comment:
This PR need to be properly rebased to avoid my changes, I'm working on rethinkdb in a separate PR #11639